### PR TITLE
Add developer's guide to upgrade dependencies and a workflow to test it

### PR DESCRIPTION
**Pull request by wanliAlex**
* **What is the current behavior?**  
n/a

* **What is the new behavior?**
- 1. Add a developer's guide to demonstrate how to change the dependencies in marqo-base with `pip-compile`. 
- 2. Add an automated test to check if the `requirements.txt` is the same as the output from `pip-compile requirements.in`

* **Have tests been run against this PR?**  
no

* **Have appropriate comments and documentation been made on this PR?**  
yes

* **Related changes in other repos** (link commit/PR here)
no

* **Other information**:
no

### CODE REVIEW

File: .github/workflows/check_requirements.yml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+  contents: read
+
+jobs:
+  tests:
+    name: Run requirements tests on multiple architectures
+    runs-on: ${{ matrix.os }}
+
+    strategy:
+      matrix:
+        os: [ubuntu-latest, ubuntu-20.04]
```
Comment (papa99do) [inline .github/workflows/check_requirements.yml:27]: why do we want to run on two different os distro?

File: README.md [hunk 2]
```diff
@@ -3,6 +3,36 @@ The dependencies and Dockerfile for creating the Marqo base image
...
+pip install pip-tools
+```
+In case you want to add a new dependency, you can add it to the `requirements.in` file and run
+```bash
+pip-compile requirements.in --output-file=requirements.txt --strip-extras
+```
+to generate the `requirements.txt` file. This file is used to install the dependencies in the Dockerfile.
+If any of the dependencies/sub-dependencies are not pinned and are updated when you run `pip-compile`, you should add
+the new version to the `requirements.in` file and run the above command again until they converge.
+We have an automated pipeline for this check. Ideally, you should not commit to the `requirements.txt` file directly.
```
Comment (papa99do) [inline README.md:20]: > Ideally, you should not commit to the `requirements.txt` file directly.
Is there any case we allow manual change of `requirements.txt` file? Show we rephrase this to be more affirmative? Like `You should not manually change 'requirements.txt' file.`

File: .github/workflows/check_requirements.yml (hunk 1 above)
Comment (wanliAlex) [reply .github/workflows/check_requirements.yml:27]: No we don't need it anymore. I will also removing running on two architectures as this is not needed either.

File: README.md (hunk 2 above)
Comment (wanliAlex) [reply README.md:20]: We need to manually change the `requirements.txt` for cross-platform dependent packages. I will rephrase it and be more specific.

File: README.md [hunk 3]
```diff
@@ -3,6 +3,37 @@ The dependencies and Dockerfile for creating the Marqo base image
...
+dependencies. For example, we have
+```text
+torch==1.12.1+cu113; platform_machine == "x86_64"
+torch==1.12.1; platform_machine == "arm64" or platform_machine == "aarch64"
+```
+The `pip-comple` tool does not support cross-platform dependencies. If you run `pip-compile` on the `requirements.in` file 
+on an `x86_64` machine, it will generate the `torch==1.12.1+cu113; platform_machine == "x86_64"` line, but remove the
+`torch==1.12.1; platform_machine == "arm64" or platform_machine == "aarch64"` line, and vice versa. 
+
+In such cases, you should manually add the missing line to the `requirements.txt` file, especially for these cross-platform
```
Comment (papa99do) [inline README.md:34]: Does this mean when adding a dependency, we'll need to run `pip-compile` on both platforms and merge the result in requirements.txt? This also makes the unit test more complex. Can we generate different  `requirements.txt` for different platforms and use the platform-specific `requirements.txt` in the Dockerfile?

File: Dockerfile [hunk 4]
```diff
@@ -25,8 +25,17 @@ RUN alternatives --set python3 /usr/bin/python3.9 && \
     curl https://bootstrap.pypa.io/get-pip.py | python3
 
 # Install pip dependencies
-COPY requirements.txt requirements.txt
-RUN pip3 install --no-cache-dir -r requirements.txt
+COPY requirements requirements
+# Install requirements based on the architecture
+RUN if [ "${TARGETPLATFORM}" = "linux/arm64" ]; then \
+      RUN pip3 install --no-cache-dir -r requirements/arm64-requirements.txt; \
```
Comment (papa99do) [inline Dockerfile:31]: remove the extra RUN command

File: requirements/requirements.in [hunk 5]
```diff
@@ -54,6 +54,36 @@ pycurl==7.45.3
...
+scikit-learn==1.3.2
+soxr==0.3.7
+termcolor==2.4.0
+tqdm==4.66.5
+transformers==4.41.2
+uvloop==0.20.0
+watchfiles==0.24.0
+websockets==13.1
+zipp==3.20.2
+yarl==1.13.1
```
Comment (papa99do) [inline requirements/requirements.in:84]: are these all transitive deps? What process did you use to decide which ones should be included?

File: Dockerfile (hunk 4 above)
Comment (wanliAlex) [reply Dockerfile:31]: removed

File: requirements/requirements.in (hunk 5 above)
Comment (wanliAlex) [reply requirements/requirements.in:84]: Yes, these are the transitive deps that are not pinned. I run `pip-compile` and compare the output with `pip freeze` on 2.13 and realize the versions of these packages are changed. So I add them in `requirements.in`

File: README.md (hunk 3 above)
Comment (wanliAlex) [reply README.md]: Yes, we will generate different `requirements.txt` file for different architectures.

**Review by papa99do — APPROVED** [review]
